File: graph_components/dfs/src/DfsDemo.go
// Go
// グラフの連結成分: DFS

package main

import (
	"fmt"
)

type EdgeInfo struct {
	Vertex string
	Weight int
}

type GraphData struct {
	data map[string][]EdgeInfo
}

func NewGraphData() *GraphData {
	return &GraphData{
		data: make(map[string][]EdgeInfo),
	}
}

func (g *GraphData) Get() map[string][]EdgeInfo {
	return g.data
}

func (g *GraphData) GetVertices() []string {
	vertices := make([]string, 0, len(g.data))
	for vertex := range g.data {
		vertices = append(vertices, vertex)
	}
	return vertices
}

func (g *GraphData) AddVertex(vertex string) bool {
	if _, exists := g.data[vertex]; !exists {
		g.data[vertex] = []EdgeInfo{}
	}
	return true
}

func (g *GraphData) AddEdge(vertex1, vertex2 string, weight int) bool {
	g.AddVertex(vertex1)
	g.AddVertex(vertex2)

	// Add or update edge from vertex1 to vertex2
	found1 := false
	for i, neighbor := range g.data[vertex1] {
		if neighbor.Vertex == vertex2 {
			g.data[vertex1][i].Weight = weight
			found1 = true
			break
		}
	}
	if !found1 {
		g.data[vertex1] = append(g.data[vertex1], EdgeInfo{Vertex: vertex2, Weight: weight})
	}

	// Add or update edge from vertex2 to vertex1
	found2 := false
	for i, neighbor := range g.data[vertex2] {
		if neighbor.Vertex == vertex1 {
			g.data[vertex2][i].Weight = weight
			found2 = true
			break
		}
	}
	if !found2 {
		g.data[vertex2] = append(g.data[vertex2], EdgeInfo{Vertex: vertex1, Weight: weight})
	}

	return true
}

func (g *GraphData) Clear() bool {
	g.data = make(map[string][]EdgeInfo)
	return true
}

func (g *GraphData) dfs(vertex string, visited map[string]bool, currentComponent *[]string) {
	visited[vertex] = true
	*currentComponent = append(*currentComponent, vertex)

	for _, neighborInfo := range g.data[vertex] {
		neighborVertex := neighborInfo.Vertex
		if !visited[neighborVertex] {
			g.dfs(neighborVertex, visited, currentComponent)
		}
	}
}

func (g *GraphData) GetConnectedComponents() [][]string {
	visited := make(map[string]bool)
	var connectedComponents [][]string

	for _, vertex := range g.GetVertices() {
		if !visited[vertex] {
			currentComponent := []string{}
			g.dfs(vertex, visited, &currentComponent)
			connectedComponents = append(connectedComponents, currentComponent)
		}
	}

	return connectedComponents
}

func main() {
	fmt.Println("Dfs TEST -----> start")

	fmt.Println("\nnew")
	graphData := NewGraphData()
	fmt.Printf("  Current data: %v\n", graphData.Get())

	fmt.Println("\nadd_edge")
	graphData.Clear()
	inputList := []struct {
		Vertex1 string
		Vertex2 string
		Weight  int
	}{
		{"A", "B", 4}, {"B", "C", 3}, {"B", "D", 2}, {"D", "A", 1}, {"A", "C", 2}, {"B", "D", 2},
	}
	for _, input := range inputList {
		fmt.Printf("  Input: (%s, %s, %d)\n", input.Vertex1, input.Vertex2, input.Weight)
		output := graphData.AddEdge(input.Vertex1, input.Vertex2, input.Weight)
		fmt.Printf("  Output: %v\n", output)
	}
	fmt.Printf("  Current data: %v\n", graphData.Get())
	fmt.Println("\nget_connected_components")
	output := graphData.GetConnectedComponents()
	fmt.Printf("  Connected components: %v\n", output)

	fmt.Println("\nadd_edge")
	graphData.Clear()
	inputList = []struct {
		Vertex1 string
		Vertex2 string
		Weight  int
	}{
		{"A", "B", 4}, {"C", "D", 4}, {"E", "F", 1}, {"F", "G", 1},
	}
	for _, input := range inputList {
		fmt.Printf("  Input: (%s, %s, %d)\n", input.Vertex1, input.Vertex2, input.Weight)
		output := graphData.AddEdge(input.Vertex1, input.Vertex2, input.Weight)
		fmt.Printf("  Output: %v\n", output)
	}
	fmt.Printf("  Current data: %v\n", graphData.Get())
	fmt.Println("\nget_connected_components")
	output = graphData.GetConnectedComponents()
	fmt.Printf("  Connected components: %v\n", output)

	fmt.Println("\nadd_edge")
	graphData.Clear()
	inputList = []struct {
		Vertex1 string
		Vertex2 string
		Weight  int
	}{
		{"A", "B", 4}, {"B", "C", 3}, {"D", "E", 5},
	}
	for _, input := range inputList {
		fmt.Printf("  Input: (%s, %s, %d)\n", input.Vertex1, input.Vertex2, input.Weight)
		output := graphData.AddEdge(input.Vertex1, input.Vertex2, input.Weight)
		fmt.Printf("  Output: %v\n", output)
	}
	fmt.Printf("  Current data: %v\n", graphData.Get())
	fmt.Println("\nget_connected_components")
	output = graphData.GetConnectedComponents()
	fmt.Printf("  Connected components: %v\n", output)

	fmt.Println("\nadd_edge")
	graphData.Clear()
	inputList = []struct {
		Vertex1 string
		Vertex2 string
		Weight  int
	}{}
	for _, input := range inputList {
		fmt.Printf("  Input: (%s, %s, %d)\n", input.Vertex1, input.Vertex2, input.Weight)
		output := graphData.AddEdge(input.Vertex1, input.Vertex2, input.Weight)
		fmt.Printf("  Output: %v\n", output)
	}
	fmt.Printf("  Current data: %v\n", graphData.Get())
	fmt.Println("\nget_connected_components")
	output = graphData.GetConnectedComponents()
	fmt.Printf("  Connected components: %v\n", output)

	fmt.Println("Dfs TEST <----- end")
}

File: graph_components/dfs/src/DfsDemo.java
// Java
// グラフの連結成分: DFS

import java.util.*;

class GraphData {
    private Map<String, List<Pair<String, Integer>>> data;

    public GraphData() {
        data = new HashMap<>();
    }

    public Map<String, List<Pair<String, Integer>>> get() {
        return data;
    }

    public List<String> getVertices() {
        return new ArrayList<>(data.keySet());
    }

    public boolean addVertex(String vertex) {
        if (!data.containsKey(vertex)) {
            data.put(vertex, new ArrayList<>());
            return true;
        }
        return true;
    }

    public boolean addEdge(String vertex1, String vertex2, int weight) {
        addVertex(vertex1);
        addVertex(vertex2);

        boolean v1v2Exists = false;
        for (int i = 0; i < data.get(vertex1).size(); i++) {
            Pair<String, Integer> neighbor = data.get(vertex1).get(i);
            if (neighbor.first.equals(vertex2)) {
                data.get(vertex1).set(i, new Pair<>(vertex2, weight));
                v1v2Exists = true;
                break;
            }
        }
        if (!v1v2Exists) {
            data.get(vertex1).add(new Pair<>(vertex2, weight));
        }

        boolean v2v1Exists = false;
        for (int i = 0; i < data.get(vertex2).size(); i++) {
            Pair<String, Integer> neighbor = data.get(vertex2).get(i);
            if (neighbor.first.equals(vertex1)) {
                data.get(vertex2).set(i, new Pair<>(vertex1, weight));
                v2v1Exists = true;
                break;
            }
        }
        if (!v2v1Exists) {
            data.get(vertex2).add(new Pair<>(vertex1, weight));
        }

        return true;
    }

    public boolean clear() {
        data.clear();
        return true;
    }

    private void dfs(String vertex, Set<String> visited, List<String> currentComponent) {
        visited.add(vertex);
        currentComponent.add(vertex);

        for (Pair<String, Integer> neighborInfo : data.getOrDefault(vertex, new ArrayList<>())) {
            String neighborVertex = neighborInfo.first;
            if (!visited.contains(neighborVertex)) {
                dfs(neighborVertex, visited, currentComponent);
            }
        }
    }

    public List<List<String>> getConnectedComponents() {
        Set<String> visited = new HashSet<>();
        List<List<String>> connectedComponents = new ArrayList<>();

        for (String vertex : getVertices()) {
            if (!visited.contains(vertex)) {
                List<String> currentComponent = new ArrayList<>();
                dfs(vertex, visited, currentComponent);
                connectedComponents.add(currentComponent);
            }
        }

        return connectedComponents;
    }

    
    // Helper classes
    public static class Pair<F, S> {
        public final F first;
        public final S second;

        public Pair(F first, S second) {
            this.first = first;
            this.second = second;
        }

        @Override
        public String toString() {
            return "(" + first + ", " + second + ")";
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Pair<?, ?> pair = (Pair<?, ?>) o;
            return Objects.equals(first, pair.first) && Objects.equals(second, pair.second);
        }

        @Override
        public int hashCode() {
            return Objects.hash(first, second);
        }
    }

    public static class Edge {
        public final String vertex1;
        public final String vertex2;
        public final int weight;

        public Edge(String vertex1, String vertex2, int weight) {
            this.vertex1 = vertex1;
            this.vertex2 = vertex2;
            this.weight = weight;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Edge edge = (Edge) o;
            return weight == edge.weight && 
                   ((Objects.equals(vertex1, edge.vertex1) && Objects.equals(vertex2, edge.vertex2)) ||
                    (Objects.equals(vertex1, edge.vertex2) && Objects.equals(vertex2, edge.vertex1)));
        }

        @Override
        public int hashCode() {
            return Objects.hash(vertex1, vertex2, weight);
        }

        @Override
        public String toString() {
            return "(" + vertex1 + ", " + vertex2 + ", " + weight + ")";
        }
    }
}

public class DfsDemo {
    public static void main(String[] args) {
        System.out.println("Dfs TEST -----> start");

        System.out.println("\nnew");
        GraphData graphData = new GraphData();
        System.out.println("  現在のデータ: " + graphData.get());

        System.out.println("\nadd_edge");
        graphData.clear();
        Object[][] inputList = {
            {"A", "B", 4}, {"B", "C", 3}, {"B", "D", 2}, 
            {"D", "A", 1}, {"A", "C", 2}, {"B", "D", 2}
        };
        for (Object[] input : inputList) {
            System.out.println("  入力値: " + Arrays.toString(input));
            boolean output = graphData.addEdge((String)input[0], (String)input[1], (Integer)input[2]);
            System.out.println("  出力値: " + output);
        }
        System.out.println("  現在のデータ: " + graphData.get());
        System.out.println("\nget_connected_components");
        List<List<String>> output = graphData.getConnectedComponents();
        System.out.println("  連結成分: " + output);

        System.out.println("\nadd_edge");
        graphData.clear();
        inputList = new Object[][] {
            {"A", "B", 4}, {"C", "D", 4}, {"E", "F", 1}, {"F", "G", 1}
        };
        for (Object[] input : inputList) {
            System.out.println("  入力値: " + Arrays.toString(input));
            boolean output2 = graphData.addEdge((String)input[0], (String)input[1], (Integer)input[2]);
            System.out.println("  出力値: " + output2);
        }
        System.out.println("  現在のデータ: " + graphData.get());
        System.out.println("\nget_connected_components");
        output = graphData.getConnectedComponents();
        System.out.println("  連結成分: " + output);

        System.out.println("\nadd_edge");
        graphData.clear();
        inputList = new Object[][] {
            {"A", "B", 4}, {"B", "C", 3}, {"D", "E", 5}
        };
        for (Object[] input : inputList) {
            System.out.println("  入力値: " + Arrays.toString(input));
            boolean output3 = graphData.addEdge((String)input[0], (String)input[1], (Integer)input[2]);
            System.out.println("  出力値: " + output3);
        }
        System.out.println("  現在のデータ: " + graphData.get());
        System.out.println("\nget_connected_components");
        output = graphData.getConnectedComponents();
        System.out.println("  連結成分: " + output);

        System.out.println("\nadd_edge");
        graphData.clear();
        inputList = new Object[0][0];
        for (Object[] input : inputList) {
            System.out.println("  入力値: " + Arrays.toString(input));
            boolean output4 = graphData.addEdge((String)input[0], (String)input[1], (Integer)input[2]);
            System.out.println("  出力値: " + output4);
        }
        System.out.println("  現在のデータ: " + graphData.get());
        System.out.println("\nget_connected_components");
        output = graphData.getConnectedComponents();
        System.out.println("  連結成分: " + output);

        System.out.println("Dfs TEST <----- end");
    }
}
